pystiche/papers/ulyanov_et_al_2016/core.py

```diff
@@ -0,0 +1,226 @@
+from typing import Union, Optional, Callable
+import torch
+from torch.utils.data import DataLoader
+from torch import nn
+import time
+from torch.optim.optimizer import Optimizer
+import pystiche
+from pystiche.optim import (
+    OptimLogger,
+    default_transformer_optim_log_fn,
+)
+from ..common_utils import batch_up_image
+from .modules import ulyanov_et_al_2016_transformer
+from .loss import (
+    UlyanovEtAl2016PerceptualLoss,
+    ulyanov_et_al_2016_perceptual_loss,
+)
+
+from .data import (
+    ulyanov_et_al_2016_style_transform,
+    ulyanov_et_al_2016_content_transform,
+    ulyanov_et_al_2016_dataset,
+    ulyanov_et_al_2016_image_loader,
+    ulyanov_et_al_2016_images,
+)
+from .utils import (
+    ulyanov_et_al_2016_preprocessor,
+    ulyanov_et_al_2016_postprocessor,
+    ulyanov_et_al_2016_optimizer,
+)
+
+__all__ = [
+    "ulyanov_et_al_2016_transformer_optim_loop",
+    "ulyanov_et_al_2016_transformer",
+    "ulyanov_et_al_2016_perceptual_loss",
+    "ulyanov_et_al_2016_dataset",
+    "ulyanov_et_al_2016_image_loader",
+    "ulyanov_et_al_2016_training",
+    "ulyanov_et_al_2016_images",
+    "ulyanov_et_al_2016_stylization",
+]
+
+
+def ulyanov_et_al_2016_transformer_optim_loop(
```

Comment:
If I'm not mistaken the whole purpose of this custom `optim_loop` is change of the learning rate (`lr`) at the end. Is that assessment correct? If yes, we should create an `epoch_optim_loop` and handle this with a `lr_scheduler`.

Reply:
In addition to the `lr`, the loss is also changed. I have already moved this to the criterion and create an `epoch_optim_loop`. However, I create an `optimizer` there, which I cannot pass to the `default_transformer_optim_loop`, because `get_optimizer` must be callable there. Would it make sense to change this so that an `optimizer` can also be transferred?

Reply:
See [my comment below](https://github.com/pmeier/pystiche/pull/88#discussion_r391459374)